plugin/evm/vm.go

```diff
@@ -620,7 +622,33 @@ func (vm *VM) Shutdown(context.Context) error {
 }
 
 // buildBlock builds a block to be wrapped by ChainState
-func (vm *VM) buildBlock(context.Context) (snowman.Block, error) {
+func (vm *VM) buildBlock(ctx context.Context) (snowman.Block, error) {
+	log.Debug("Trying to build block without context.")
+	return vm.buildBlockWithContext(ctx, nil)
+}
+
+func (vm *VM) buildBlockWithContext(ctx context.Context, proposerVMBlockCtx *block.Context) (snowman.Block, error) {
+	if proposerVMBlockCtx != nil {
+		log.Debug("Building block with context", "pChainBlockHeight", proposerVMBlockCtx.PChainHeight)
+	} else {
+		log.Debug("Attempting to build block with nil context")
+	}
+
+	// Get the pending transactions from the pool.
+	pending := vm.txPool.Pending(true)
+
+	// Verify any transaction predicates with the given block context.
+	// Remove any transactions from the pool from addresses that submitted transactions with invalid predicates.
+	for sender, txs := range pending {
+		if invalidIndex, err := vm.checkTransactionPredicates(txs, proposerVMBlockCtx); err != nil {
+			log.Debug("Removing transactions from sender of transaction with invalid predicate.", "sender", sender.Hex())
+			delete(pending, sender)
+			for i := invalidIndex; i < len(txs); i++ {
+				vm.txPool.RemoveTx(txs[i].Hash())
+			}
```

Comment:
can we add a comment why these txs are removed?

Reply:
I believe these transactions should be passed in to `vm.miner.GenerateBlock(txs)` as an argument after filtering out any transactions that fail the predicate.

Reply:
The call to `PendingWithPredicates` removes txs that fail the predicate from the pending pool. 

> I believe these transactions should be passed in to vm.miner.GenerateBlock(txs) as an argument after filtering out any transactions that fail the predicate.

Not sure what change you are suggesting here; `GenerateBlock` does not accept any parameters.

Reply:
I am suggesting to add a parameter to `GenerateBlock` so that it takes transactions as a parameter instead of calling `Pending(true)` inside of the miner.

Reply:
Ah gotcha, thanks for clarifying. Made this change now.